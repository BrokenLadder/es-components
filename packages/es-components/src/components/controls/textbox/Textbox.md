The Textbox component will accept typical input attributes as props such as onChange, onBlur, value, placeholder, etc.

```
const Control = require('../Control').default;

<>
  <Control>
    <Label for="stacked">Stacked</Label>
    <Textbox id="stacked" />
  </Control>

  <Control orientation="inline">
    <Label for="inline">Inline</Label>
    <Textbox id="inline" />
  </Control>
</>
```

Pass an `onChange` function to execute any time the input box value changes. This function will have the current value of the input box passed to it. The following is a simple controlled example, where state is handled outside the component. (See [Controlled Components](https://facebook.github.io/react/docs/forms.html#controlled-components) and [Uncontrolled Components](https://facebook.github.io/react/docs/uncontrolled-components.html).)

```
const Control = require('../Control').default;

class TextboxExample extends React.Component {
  constructor() {
    this.state = { value: "" };
    this.handleOnTextChange = this.handleOnTextChange.bind(this);
  }

  handleOnTextChange(event) {
    console.log(event.target.value);
    this.setState({ value: event.target.value });
  }

  render() {
    return (
      <Control>
        <Label for="controlled-example">Controlled Example</Label>
        <Textbox
          id="controlled-example"
          value={state.value}
          onChange={this.handleOnTextChange}
        />
      </Control>
    )
  }
}

<TextboxExample />
```

Pass an `onBlur` function to execute when the text box loses focus. This function will have the current value of the input passed to it. To provide a default value in an uncontrolled `Textbox` component, use `defaultValue`.

```
const Control = require('../Control').default;

function handleOnBlur(event) {
  if(event.target.value.length > 0) {
    confirm(`You entered ${event.target.value}. Is that correct?`);
  }
}

<Control>
  <Label>First name</Label>
  <Textbox
    defaultValue="Default"
    labelText="First name"
    onBlur={handleOnBlur}
  />
</Control>
```

### Validation states

Validation states are driven from the `Control` component. This is what they look like for Textboxes.

```
const Control = require('../Control').default;
const AdditionalHelp = require('../AdditionalHelp').default;

<>
  <Control validationState="success">
    <Label for="success">Success</Label>
    <Textbox id="success" aria-describedby="success-help" />
    <AdditionalHelp id="success-help">When validationState is set to success</AdditionalHelp>
  </Control>

  <Control validationState="warning">
    <Label for="warning">Warning</Label>
    <Textbox id="warning" aria-describedby="warning-help" />
    <AdditionalHelp id="warning-help">When validationState is set to warning</AdditionalHelp>
  </Control>

  <Control validationState="danger">
    <Label for="danger">Danger</Label>
    <Textbox id="danger" aria-describedby="danger-help" />
    <AdditionalHelp id="danger-help">When validationState is set to danger</AdditionalHelp>
  </Control>
</>
```

### Appending and Prepending text

Provide an `appendText` or `prependText` prop for appending and prepending inputs. Validation state colorings will also be applied.

```
const Control = require('../Control').default;

<>
  <Control>
    <Label for="appended">Appended</Label>
    <Textbox
      id="appended"
      appendIconName="calendar"
    />
  </Control>

  <Control>
    <Label for="prepended">Prepended</Label>
    <Textbox
      id="prepended"
      prependIconName="phone"
    />
  </Control>

  <Control>
    <Label for="both">Appended and prepended</Label>
    <Textbox
      id="both"
      appendIconName="user"
      prependIconName="phone"
    />
  </Control>

  <Control validationState="success">
    <Label for="appended">Appended</Label>
    <Textbox
      id="appended"
      appendIconName="calendar"
    />
  </Control>

  <Control validationState="warning">
    <Label for="prepended">Prepended</Label>
    <Textbox
      id="prepended"
      prependIconName="phone"
    />
  </Control>

  <Control validationState="danger">
    <Label for="both">Appended and prepended</Label>
    <Textbox
      id="both"
      appendIconName="user"
      prependIconName="phone"
    />
  </Control>
</>
```

### Text masks

```
const Control = require('../Control').default;

<>
  <Control>
    <Label for="dollar">US Dollar</Label>
    <Textbox id="dollar" maskType="dollar" />
  </Control>

  <Control>
    <Label for="phone-number">Phone number</Label>
    <Textbox 
      id="phone-number"
      maskType="phone"
      title="Providing a title will override the default mask title text"
    />
  </Control>

  <Control>
    <Label for="ssn">Social Security Number</Label>
    <Textbox id="ssn" maskType="ssnum" />
  </Control>

  <Control>
    <Label for="zip">Zip code</Label>
    <Textbox id="zip" maskType="zip" />
  </Control>

  <Control>
    <Label for="date">Date</Label>
    <Textbox id="date" maskType="date" placeholder="mm/dd/yyyy" />
  </Control>
</>
```

### Custom masks

Create your own text mask using the structure documented [here](https://github.com/text-mask/text-mask/blob/master/componentDocumentation.md#text-mask-documentation).

```
const Control = require('../Control').default;

const mask = {
    mask: [/[A-Za-z]/, /[A-Za-z]/, /[A-Za-z]/, '-', /[A-Za-z]/, /[A-Za-z]/, /[A-Za-z]/],
    showMask: true,
    keepCharPositions: false,
    placeholderChar: '_'
  };

<Control>
  <Label for="mask">Enter 6 letters (no numbers)</Label>
  <Textbox id="mask" maskType="custom" customMask={mask} />
</Control>
```

### Additional props

Other typical input attributes passed to `Textbox` are allowed. The react `autoFocus` property is also accepted.

```
const Control = require('../Control').default;

<>
  <Control>
    <Label for="disabled">Disabled</Label>
    <Textbox
      id="disabled"
      defaultValue="John Doe"
      disabled
    />
  </Control>

  <Control>
    <Label for="readonly">Readonly</Label>
    <Textbox
      id="readonly"
      defaultValue="Jane Doe"
      readOnly
    />
  </Control>

  <Control>
    <Label for="placeholder">Placeholder</Label>
    <Textbox
      id="placeholder"
      placeholder="Placeholder is one of many input attributes"
    />
  </Control>
</>
```
