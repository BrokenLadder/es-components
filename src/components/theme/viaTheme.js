const white = '#fff';
const gray1 = '#f4f4f4';
const gray2 = '#ecedee';
const gray3 = '#e4e4e4';
const gray4 = '#d8d8d8';
const gray5 = '#c5c5c5';
const gray6 = '#979797';
// const gray7 = '#707070';
const gray8 = '#585858';
const gray9 = '#444';
const black = '#000';

const grayLightest = gray1;
const grayLighter = gray2;
const grayLight = gray3;
const gray = gray5;
const grayDark = gray6;
const grayDarker = gray8;
const grayDarkest = gray9;

const dflt = '#d8d8d8';
const dfltHover = '#bfbfbf';
const primary = '#007fa7';
const primaryHover = '#005874';
const info = '#069';
const infoHover = '#004466';
const success = '#298544';
const successHover = '#1d5e30';
const warning = '#b35f00';
const warningHover = '#804400';
const danger = '#c00';
const dangerHover = '#990000';
const advisor = warning;

const boxShadowLight = 'rgba(0, 0, 0, 0.075)';
const boxShadowDark = 'rgba(0, 0, 0, 0.5)';

const inputBorder = gray4;
const inputFocus = '#83bffc';
const inputBoxShadow = 'rgba(102, 175, 233, 0.6)';

// Datepicker colors
const navArrow = gray;
const navArrowHover = grayDark;
const selected = '#216ba5';
const hover = '#1d5d90';
const keyboard = '#2a87d0';
const inRange = 'rgba(33, 107, 165, 0.5)';
const highlight = '#3dcc4a';
const highlightHover = '#32be3f';

const theme = {
  colors: {
    black,
    grayDarkest,
    grayDarker,
    grayDark,
    gray,
    grayLight,
    grayLighter,
    grayLightest,
    white,
    dflt,
    dfltHover,
    primary,
    primaryHover,
    success,
    successHover,
    info,
    infoHover,
    warning,
    warningHover,
    danger,
    dangerHover,
    advisor,
    boxShadowLight,
    boxShadowDark,
    inputFocus
  },
  datepickerColors: {
    navArrow,
    navArrowHover,
    selected,
    hover,
    keyboard,
    inRange,
    highlight,
    highlightHover
  },
  buttonStyles: {
    buttonsNormal: {
      primary: {
        bgColor: primary,
        textColor: white,
        hoverBgColor: primaryHover,
        hoverTextColor: white,
        activeBgColor: primaryHover,
        activeTextColor: white,
        boxShadowColor: primaryHover
      },
      secondary: {
        bgColor: dflt,
        textColor: white,
        hoverBgColor: dfltHover,
        hoverTextColor: white,
        activeBgColor: dfltHover,
        activeTextColor: white,
        boxShadowColor: dfltHover
      },
      success: {
        bgColor: success,
        textColor: white,
        hoverBgColor: successHover,
        hoverTextColor: white,
        activeBgColor: successHover,
        activeTextColor: white,
        boxShadowColor: successHover
      },
      info: {
        bgColor: info,
        textColor: white,
        hoverBgColor: infoHover,
        hoverTextColor: white,
        activeBgColor: infoHover,
        activeTextColor: white,
        boxShadowColor: infoHover
      },
      warning: {
        bgColor: warning,
        textColor: white,
        hoverBgColor: warningHover,
        hoverTextColor: white,
        activeBgColor: warningHover,
        activeTextColor: white,
        boxShadowColor: warningHover
      },
      danger: {
        bgColor: danger,
        textColor: white,
        hoverBgColor: dangerHover,
        hoverTextColor: white,
        activeBgColor: dangerHover,
        activeTextColor: white,
        boxShadowColor: dangerHover
      }
    },
    buttonsAlt: {
      primary: {
        bgColor: white,
        textColor: primary,
        hoverBgColor: primary,
        hoverTextColor: white,
        activeBgColor: primaryHover,
        activeTextColor: white,
        borderColor: primary
      },
      gray: {
        bgColor: white,
        textColor: '#63666a',
        hoverBgColor: '#63666a',
        hoverTextColor: white,
        activeBgColor: '#4a4d50',
        activeTextColor: white,
        borderColor: '#63666a'
      },
      magenta: {
        bgColor: white,
        textColor: '#c110a0',
        hoverBgColor: '#c110a0',
        hoverTextColor: white,
        activeBgColor: '#920C79',
        activeTextColor: white,
        borderColor: '#c110a0'
      },
      violet: {
        bgColor: white,
        textColor: '#702082',
        hoverBgColor: '#702082',
        hoverTextColor: white,
        activeBgColor: '#4d1659',
        activeTextColor: white,
        borderColor: '#702082'
      }
    }
  },
  // buttonSizes should always have default, lg, sm, xs
  buttonSizes: {
    lg: {
      borderRadius: '3px',
      fontSize: '26px',
      padding: '8px 20px 5px'
    },
    default: {
      borderRadius: '2px',
      fontSize: '18px',
      padding: '5px 15px 4px'
    },
    sm: {
      borderRadius: '2px',
      fontSize: '14px',
      padding: '4px 10px 3px'
    },
    xs: {
      borderRadius: '2px',
      fontSize: '12px',
      padding: '3px 10px 2px'
    }
  },
  validationIconName: {
    success: 'ok',
    warning: 'warning-sign',
    danger: 'remove'
  },
  validationTextColor: {
    success,
    warning: warningHover,
    danger
  },
  validationInputColor: {
    success: {
      borderColor: success,
      boxShadow: `inset 0 0 5px ${success}`,
      focusBorderColor: success,
      focusBoxShadow: `0 0 6px ${success}`
    },
    warning: {
      borderColor: warning,
      boxShadow: `inset 0 0 5px ${warning}`,
      focusBorderColor: warning,
      focusBoxShadow: `0 0 6px ${warning}`
    },
    danger: {
      borderColor: danger,
      boxShadow: `inset 0 0 5px ${danger}`,
      focusBorderColor: danger,
      focusBoxShadow: `0 0 6px ${danger}`
    },
    default: {
      borderColor: inputBorder,
      boxShadow: `inset 0 1px 1px ${boxShadowLight}`,
      focusBorderColor: inputFocus,
      focusBoxShadow: `0 0 8px ${inputBoxShadow}`
    }
  },
  sizes: {
    baseFontSize: 18,
    baseLineHeight: 24
  },
  screenSize: {
    retina: '320px',
    phone: '480px',
    tablet: '768px',
    desktop: '992px',
    widescreen: '1200px'
  }
};

export default theme;
