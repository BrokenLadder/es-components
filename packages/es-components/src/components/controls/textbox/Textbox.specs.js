/* eslint-env jest */

import React from 'react';
import { render, fireEvent, cleanup } from 'react-testing-library';
import { ThemeProvider } from 'styled-components';
import viaTheme from 'es-components-via-theme';

import Textbox from './Textbox';

beforeEach(cleanup);

const buildTextbox = props => (
  <ThemeProvider theme={viaTheme}>
    <Textbox labelText="Text" {...props} />
  </ThemeProvider>
);

it('includes additional text when labelSuffix is provided', () => {
  const props = {
    onChange: jest.fn(),
    value: 'testvalue',
    labelSuffix: <span>Optional</span>
  };
  const { queryByText } = render(buildTextbox(props));
  expect(queryByText('Optional')).not.toBeNull();
});

it('executes the handleOnChange function when text is changed', () => {
  const props = {
    onChange: jest.fn()
  };
  const { getByLabelText } = render(buildTextbox(props));

  fireEvent.change(getByLabelText('Text'), {
    target: { value: '112' }
  });
  expect(props.onChange).toHaveBeenCalled();
});

it('executes handleOnBlur when input focus is lost', () => {
  const props = {
    onBlur: jest.fn()
  };
  const { getByLabelText } = render(buildTextbox(props));

  fireEvent.blur(getByLabelText('Text'));
  expect(props.onBlur).toHaveBeenCalled();
});

it('renders addons when provided', () => {
  function getNumberOfIcons(props) {
    const { container } = render(buildTextbox(props));
    return container.querySelectorAll('i').length;
  }

  expect(getNumberOfIcons({ prependIconName: 'prepend' })).toBe(1);
  expect(getNumberOfIcons({ appendIconName: 'append' })).toBe(1);
  expect(getNumberOfIcons({ prependIconName: 'prepend', appendIconName: 'append' })).toBe(2);
});

it('applies mask to changed value', () => {
  const props = {
    maskType: 'ssnum'
  };
  const { getByLabelText } = render(buildTextbox(props));

  fireEvent.change(getByLabelText('Text'), { target: {
    value: '123452323'
  }});

  expect(getByLabelText('Text').value).toBe('123-45-2323');
});

it('updates mask properly', () => {
  const props = {
    maskType: 'ssnum'
  };
  const { getByLabelText } = render(buildTextbox(props));

  fireEvent.change(getByLabelText('Text'), { target: {
    value: '8'
  }});

  expect(getByLabelText('Text').value).toBe('8\u2000\u2000-\u2000\u2000-\u2000\u2000\u2000\u2000')
});
