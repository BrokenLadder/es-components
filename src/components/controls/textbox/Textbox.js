import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { noop, isNil } from 'lodash';

import Icon from '../../base/icons/Icon';
import { sizes } from '../../theme';
import { LabelText, InputBase } from '../BaseControls';
import Label from '../Label';
import Addon from './Addon';
import getValidationStateVariables from '../getValidationStateVariables';
import getAddonType from './getAddonType';
import genId from '../../util/generateAlphaName';

const TextBoxLabel = styled(Label)`
  flex-basis: 50%;
`;

function getBorderRadius(addonType) {
  switch (addonType) {
    case 'append':
      return '2px 0 0 2px';
    case 'prepend':
      return '0 2px 2px 0';
    case 'both':
      return '0';
    default:
      return '2px';
  }
}

const StyledText = styled(InputBase)`
  border-radius: ${props => getBorderRadius(props.addonType)};
`;

const AdditionalHelpContent = styled.div`
  font-size: ${sizes.baseFontSize}px;
  font-weight: normal;
  margin: 10px 0 20px 0;
  text-transform: none;
`;

const TextWrapper = styled.div`
  flex: auto;
  margin-right: ${props => (props.includeMargin ? '20px' : '0')};
  position: relative;
`;

const TextIcon = styled(Icon)`
  font-weight: normal;
  position: absolute;
  right: 9px;
  top: 9px;
`;

const InputWrapper = styled.div`
  display: flex;
  flex: auto;
`;

class Textbox extends React.Component {
  static propTypes = {
    labelText: PropTypes.string.isRequired,
    /** The name of the input */
    name: PropTypes.string,
    /** Display label inline with text box */
    inline: PropTypes.bool,
    /** Function to execute when text box value changes */
    handleOnChange: PropTypes.func,
    /** Function to execute when text box loses focus */
    handleFocusLost: PropTypes.func,
    /** Content to display underneath the text box */
    additionalHelpContent: PropTypes.node,
    /** Display label and text with contextual state colorings */
    validationState: PropTypes.oneOf(['success', 'warning', 'danger']),
    /** Content to prepend input box with */
    prependContent: PropTypes.node,
    /** Content to append to input box */
    appendContent: PropTypes.node,
    /** Default value of the textbox */
    initialValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
  };

  static defaultProps = {
    inline: false,
    handleOnChange: noop,
    handleFocusLost: noop,
    initialValue: ''
  };

  constructor(props) {
    super(props);

    this.state = {
      currentValue: isNil(props.initialValue) ? '' : props.initialValue
    };

    this.setTextboxRef = this.setTextboxRef.bind(this);
    this.handleOnTextChanged = this.handleOnTextChanged.bind(this);
    this.handleOnTextFocusLost = this.handleOnTextFocusLost.bind(this);
  }

  componentWillReceiveProps({ initialValue }) {
    this.setState(() => ({
      currentValue: isNil(initialValue) ? '' : initialValue
    }));
  }

  setTextboxRef(ref, type) {
    if (type) {
      this[`${type}Ref`] = ref;
    }
  }

  executeHandler(event, handlerName) {
    const { value } = event.target;
    const handler = this.props[handlerName];
    this.setState({ currentValue: value }, () => handler(value));
  }

  handleOnTextChanged(event) {
    this.executeHandler(event, 'handleOnChange');
  }

  handleOnTextFocusLost(event) {
    this.executeHandler(event, 'handleFocusLost');
  }

  renderAddon(type, addonContent, inputVariables, shouldRenderAddon, inline) {
    if (shouldRenderAddon) {
      return (
        <Addon
          ref={addon => this.setTextboxRef(addon, type)}
          className={type}
          backgroundColor={inputVariables.addonBackgroundColor}
          borderColor={inputVariables.borderColor}
          inline={inline}
        >
          {addonContent}
        </Addon>
      );
    }
    return null;
  }

  render() {
    const {
      labelText,
      name,
      inline,
      additionalHelpContent,
      validationState,
      prependContent,
      appendContent,
      ...additionalTextProps
    } = this.props;

    const inputVariables = getValidationStateVariables(validationState);

    const hasHelpContent = additionalHelpContent !== undefined;

    const textboxId = genId();
    const helpId = hasHelpContent ? genId() : null;

    const additionalHelp =
      hasHelpContent &&
      <AdditionalHelpContent id={helpId} className="textbox__help">
        {additionalHelpContent}
      </AdditionalHelpContent>;

    const hasPrependedText = prependContent !== undefined;
    const hasAppendedText = appendContent !== undefined;

    const addonType = getAddonType(hasPrependedText, hasAppendedText);
    const hasNoAddon = addonType === null;

    const icon = inputVariables.icon !== undefined
      ? <TextIcon name={inputVariables.icon} size={20} />
      : null;

    const inputName = name || labelText.replace(/\s+/g, '');

    return (
      <TextBoxLabel
        htmlFor={textboxId}
        color={inputVariables.foregroundColor}
        inline={inline}
      >
        <LabelText inline={inline}>{labelText}</LabelText>
        <InputWrapper>
          {this.renderAddon(
            'prepend',
            prependContent,
            inputVariables,
            hasPrependedText
          )}
          <TextWrapper includeMargin={hasNoAddon && inline}>
            <StyledText
              id={textboxId}
              ref={ref => this.setTextboxRef(ref, 'textbox')}
              addonType={addonType}
              type="text"
              name={inputName}
              onChange={this.handleOnTextChanged}
              onBlur={this.handleOnTextFocusLost}
              aria-describedby={helpId}
              value={this.state.currentValue}
              {...inputVariables}
              {...additionalTextProps}
              innerRef={this.setTextboxRef}
            />
            {icon}
          </TextWrapper>
          {this.renderAddon(
            'append',
            appendContent,
            inputVariables,
            hasAppendedText,
            inline
          )}
        </InputWrapper>
        {additionalHelp}
      </TextBoxLabel>
    );
  }
}

export default Textbox;
