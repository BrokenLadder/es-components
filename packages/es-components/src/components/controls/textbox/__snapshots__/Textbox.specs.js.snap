// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textbox component generates id if no id is provided 1`] = `
<label
  className="es-textbox sc-EHOje YOrxh sc-bwzfXH chwbqs"
  htmlFor="abcdef"
>
  <span
    className="es-textbox__label sc-bZQynM bVIbJX sc-htpNat iykjZL"
  >
    Text
  </span>
  <div
    className="es-textbox__wrapper sc-jzJRlG dVsVQw"
  >
    <input
      className="es-textbox__input--default sc-iwsKbI gdXscJ sc-bxivhb kvEbwS"
      id="abcdef"
      onChange={[MockFunction]}
      type="text"
      value="testvalue"
    />
  </div>
</label>
`;

exports[`Textbox component includes additional text when labelSuffix is provided 1`] = `
<label
  className="es-textbox sc-EHOje YOrxh sc-bwzfXH chwbqs"
  htmlFor="abcdef"
>
  <span
    className="es-textbox__label sc-bZQynM bVIbJX sc-htpNat iykjZL"
  >
    Text
    <span
      className="sc-gzVnrw psQJc"
    >
      <span>
        Optional
      </span>
    </span>
  </span>
  <div
    className="es-textbox__wrapper sc-jzJRlG dVsVQw"
  >
    <input
      className="es-textbox__input--default sc-iwsKbI gdXscJ sc-bxivhb kvEbwS"
      id="abcdef"
      onChange={[MockFunction]}
      type="text"
      value="testvalue"
    />
  </div>
</label>
`;

exports[`Textbox component renders as expected 1`] = `
<label
  className="es-textbox sc-EHOje YOrxh sc-bwzfXH chwbqs"
  htmlFor="test-id"
>
  <span
    className="es-textbox__label sc-bZQynM bVIbJX sc-htpNat iykjZL"
  >
    Text
  </span>
  <div
    className="es-textbox__wrapper sc-jzJRlG dVsVQw"
  >
    <input
      className="es-textbox__input--default sc-iwsKbI gdXscJ sc-bxivhb kvEbwS"
      id="test-id"
      name="test-textbox"
      onChange={[MockFunction]}
      type="text"
      value="testvalue"
    />
  </div>
</label>
`;
